### DISCUSSION TITLE
On Subinterpreters and resetting the interpreter state

### DISCUSSION THREAD

**Discussion by simonegiacomelli** [pyodide/pyodide#4849] [Q&A]
Hey everybody,

I'm implementing a web framework using Pyodide (a [talk](https://www.youtube.com/watch?v=zGYVIbeUWUg) about it in PyCon Es 2023 and a [demo](https://github.com/www-py/wwwpy-demo))

I would like to implement some hot reload feature (now I just hit the browser refresh button).
I think one option is to use subinterpreters. (other ideas to reset the interpreter state in the browser?).

My current issue is that `import js` in a subinterpreter fails.
```
import _xxsubinterpreters as subinterpreters
subinterpreters.run_string(subinterpreters.create(), 'import js\n')
```
With the following error: 
`_xxsubinterpreters.RunFailedError: <class 'ModuleNotFoundError'>: No module named 'js'`

I know little about subinterpreter and noticed that I'm importing a private module... Is this intended or by design?

My overall goal is to switch from a development cycle where I refresh the browser to get the changes to reset the state of the interpreter (Pyodide) in the browser memory and reload from there.

Another option would be to use `importlib` to reload modules but I think it is not a clean and easy solution...

Any ideas?
Thanks!

**Comment by hoodmane** [comment]
> I know little about subinterpreter and noticed that I'm importing a private module... Is this intended or by design?

It's not intended that js module doesn't work in a subinterpreter. The use case makes a lot of sense -- it is desirable to have some reasonably lightweight approach to isolation. I'm not sure whether subinterpreters are the best fit for this, but if they are then we should make them work.

cc @ericsnowcurrently

**Reply by simonegiacomelli** [reply]
I don't know any better alternative at this time... 
Is there some mechanism to reset Pyodide state from javascript?

I briefly used `importlib` in the past but I believe it would be a worst fit than subinterpreters: it looks like there are many pitfalls reloading a bunch of modules.

**Comment by ryanking13** [comment]
I don't know much about subinterpreters, but doesn't it require threading? I am not sure if subinterpreters will work well with Pyodide without threading support.

**Reply by hoodmane** [reply]
I think it doesn't require threading, it's just primarily intended to make threading work better. But I could be wrong.

**Reply by simonegiacomelli** [reply]
I can confirm that the following async code is working perfectly inside a Pyodide subinterpreter.

```python
import asyncio
from datetime import datetime
from pathlib import Path

file_path = Path('/log.txt')
file_path.write_text('init\n')


def log(msg):
    file_path.write_text(file_path.read_text() + msg + '\n')

async def write_to_file():
    while True:
        log(datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f'))
        await asyncio.sleep(1)

loop = asyncio.get_event_loop()
loop.run_until_complete(write_to_file())
```

I know that it works because the file is being updated with the datetime.

If someone can give me some pointers where to look I could go through the code and try to understand what's going on with the `import js` failing...

**Reply by ryanking13** [reply]
Interesting. I thought each subinterpreter should have a separate thread because I often saw [this kind of image](https://github.com/pyodide/pyodide/assets/24893111/81897dd5-bdf5-4576-beb3-71720725df76), but I was probably misunderstanding the concept of subinterpreter.

> If someone can give me some pointers where to look I could go through the code and try to understand what's going on with the import js failing...

We define our [importhook here](https://github.com/pyodide/pyodide/blob/d2d8865a04ddc995de2af21b1cff42b3dc0efae6/src/py/_pyodide/_importhook.py#L12), and register them [here](https://github.com/pyodide/pyodide/blob/d2d8865a04ddc995de2af21b1cff42b3dc0efae6/src/js/api.ts#L781-L783) when initializing pyodide. I think it doesn't work in subinterpreters because those states (sys.meta_path) are not shared between interpreters.

**Reply by hoodmane** [reply]
I think we also have the problem that we define many of our C types as static types and they should be switched to heap types.

**Reply by simonegiacomelli** [reply]
@ryanking13 Thanks for the links; I read through the code and I think I got the raw idea on how the module js works and is made available.

Pep 554 and 734 are both on subinterpreter and one [example](https://peps.python.org/pep-0554/#run-isolated-code-in-current-os-thread) is about using the host thread.

I did many attempts but as for now I wasn't able to find a way to infiltrate the js module inside the subinterpreter.
I think of particular interest is the section [shared data](https://peps.python.org/pep-0554/#shared-data) of pepe 554.

I looked at it and noticed `_state.save_state()` and `_state.restore_state(...)`; I tried to use these but after restoring an early state I still had some used module in memory... maybe I used them improperly?

If you have ideas or pointers, I'm all ears.

**Reply by ryanking13** [reply]
Hmm, I am not sure if it is something we can do in Pyodide side, In python 3.12, there is no user friendly APIs for subinterpreters so maybe we need to wait for upstream support on sharing data?

Another possible way would be exposing some APIs like `register_js_finder`, so users can register sys.meta_path in subinterpreters by calling that method.

**Reply by ericsnowcurrently** [reply]
> I think it doesn't work in subinterpreters because those states (sys.meta_path) are not shared between interpreters.

Correct.  Each interpreter is completely isolated in that regard.  They each run the same initialization code, so the initial state should be the same.  If after that your program modifies things like `sys.meta_path`, those changes will not propagate into subinterpreters.  You would have to re-run the relevant code from your program in each subinterpreter.

> I did many attempts but as for now I wasn't able to find a way to infiltrate the js module inside the subinterpreter.

I'm guessing that some code needs to run in the subinterpreter to set up `sys.path` or `sys.meta_path` or something.

> Hmm, I am not sure if it is something we can do in Pyodide side, In python 3.12, there is no user friendly APIs for subinterpreters so maybe we need to wait for upstream support on sharing data?

Yeah, that's something I'm working on.  I've published the PEP 734 implementation as https://pypi.org/project/interpreters-pep-734/, with plans for some improvements in the near future, but currently it's only for 3.13.  I'm planning on supporting 3.12 with that module if it's possible.  I made an attempt with https://pypi.org/project/interpreters-3-12/ but had to prioritize other things so that module is out of date.
